src/traces/sankey/defaults.js

```diff
@@ -0,0 +1,99 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Lib = require('../../lib');
+var attributes = require('./attributes');
+var colors = require('../../components/color/attributes').defaults;
+var Color = require('../../components/color');
+var tinycolor = require('tinycolor2');
+var tarjan = require('strongly-connected-components');
+
+function circularityPresent(nodeList, sources, targets) {
+
+    var nodes = nodeList.map(function() {return [];});
+
+    for(var i = 0; i < Math.min(sources.length, targets.length); i++) {
```

Comment:
I think I already commented on this, but can we move the computations that scale as the length of the data arrays in `calc.js`?